### ISSUE TITLE
Проблема с bemhtml после обновления до bem-core 2.7.0

### ISSUE THREAD

**Issue by Guria** [bem/bem-forum-content-ru#600]
```js
block('kg-appbar')(
    js()(true),
    tag()('header'),
    elem('title').tag()('h1'),
    elem('panel')(
        tag()('nav'),
        def()(function(){
            return applyCtx({
                elem: 'placeholder',
                content: this.ctx
            });
        })
    ),
);
```
с 2.6.0 работало. после обновления:
```
RangeError: Maximum call stack size exceeded
    at applyc (bundle.js:443)
    at __$b37 (bundle.js:3822)
    at __$g0 (bundle.js:4875)
    at applyc (bundle.js:460)
    at __$b89 (bundle.js:4225)
    at applyc (bundle.js:573)
    at __$b19 (bundle.js:3333)
    at __$g0 (bundle.js:4688)
    at applyc (bundle.js:460)
    at __$b89 (bundle.js:4225)
```

**Comment by Guria** [comment]
В [песочнице](https://rozhdestvenskiy.ru/dev/bemhtml/) работает, но не знаю какая там версия bem-core

**Comment by Guria** [comment]
bemjson для тестов
```js
{
    "block": "kg-appbar",
    "logo": {
        "block": "logo",
        "mods": {},
        "content": [{"block": "logo", "elem": "img", "mods": {}}, {"tag": "span", "content": "Сотрудник"}]
    },
    "controls": [{"block": "user", "mods": {}}],
    "mods": {},
    "content": [{
        "elem": "panel",
        "content": [{
            "elem": "logo",
            "content": {
                "block": "logo",
                "mods": {},
                "content": [{"block": "logo", "elem": "img", "mods": {}}, {"tag": "span", "content": "Сотрудник"}]
            }
        }, {"elem": "app-title"}, {"elem": "controls", "content": [{"block": "user", "mods": {}}]}]
    }, {"elem": "progressbar", "mods": {"indeterminate": true}}, [{}]]
}
```

**Comment by Guria** [comment]
Воспроизводится на последнем project-stub https://github.com/Guria/project-stub/tree/bemhtml-bug

**Comment by tadatuta** [comment]
@Guria предположу, что такое поведение появилось не после перехода на новую версию bem-core, а после перехода на мажорную версию `enb-bemxjst`?

Там среди прочих изменений новая версия `bem-xjst`, которая научилась рекурсивно выполнять `applyNext`/`applyCtx`. Поэтому, для случая, когда необходимо завернуть текущий контекст, следует использовать `wrap()`: https://github.com/bem/project-stub/commit/3b17073a1ef4980de908f75d767d4283b75df323

**Comment by Guria** [comment]
@tadatuta в данный момент изменилась только версия bem-components и bem-core.
npm зависимости и конфиг enb не менялись. Откат до 2.6.0 устраняет проблему.

**Comment by Guria** [comment]
https://ru.bem.info/technology/bemhtml/v2/reference/#Оборачивание-блока-в-другой-блок
ни слова про wrap

**Comment by tadatuta** [comment]
Да, это я ошибся. Изменение посчитали исправлением бага: https://ru.bem.info/libs/bem-core/v2.7.0/changelog/#В-релиз-вошли-следующие-исправления-ошибок («Исправлена ошибка в i-bem.bemhtml, из-за которой неверно интерпретировались вложенные вызовы applyNext (b1dc50c).»).

В референс про `wrap()` добавим, завел [задачу](https://github.com/bem/bem-core/issues/1219). Спасибо!

**Comment by Guria** [comment]
@tadatuta мне кажется или это тянет на мажорный апдейт по семвер?

**Comment by Guria** [comment]
Кстати ещё для @sbmaxx задачка - обновить bem-core в песочнице

**Comment by tadatuta** [comment]
@Guria наверное да, но поздно пить боржоми :(
Изменение вошло в минор именно потому, что его рассмотрели как багфикс.

**Comment by Guria** [comment]
Грустно. Это стоит как минимум подробно осветить в ченджлоге.

**Comment by sbmaxx** [comment]
> Кстати ещё для @sbmaxx задачка - обновить bem-core в песочнице

попробую вечерком обновить

**Comment by apsavin** [comment]
[Считается](http://semver.org/), что если обнаружено изменение, которое ломает обратную совместимость, то нужно выкатить патч с ревертом. А потом уже выкатить мажорный апдейт, если изменение полезное.

Но я ни к чему не призываю, нет.

**Comment by apsavin** [comment]
Как узнают об этой проблеме все те пользователи bemhtml, которые не следят за этим репо?

**Comment by tadatuta** [comment]
@h4 если падает с чем-то похожим на
```
RangeError: Maximum call stack size exceeded
    at applyc (bundle.js:443)
    at __$b37 (bundle.js:3822)
    at __$g0 (bundle.js:4875)
    at applyc (bundle.js:460)
    at __$b89 (bundle.js:4225)
    at applyc (bundle.js:573)
    at __$b19 (bundle.js:3333)
    at __$g0 (bundle.js:4688)
    at applyc (bundle.js:460)
    at __$b89 (bundle.js:4225)
```
то это определенно `xjst`, т.е. BEMHTML либо BEMTREE, других вариантов нет.

**Comment by h4** [comment]
@tadatuta видимо что-то другое. Вот такая ошибка
```
InternalError in / too much recursion
```
Стек вызовов потеряли, к сожалению.

**Comment by sbmaxx** [comment]
> @h4 если падает с чем-то похожим на

даёшь `bem-xjst-2` с нормальной отладкой :)

**Comment by h4** [comment]
@sbmaxx чтобы меня окончательно в дурку забрали?

**Comment by sbmaxx** [comment]
@h4 почему?

**Comment by zxqfox** [comment]
@h4 `grep 'InternalError in ' ./node_modules -r` ?

**Comment by tadatuta** [comment]
@zxqfox это нативное браузерное сообщение

**Comment by zxqfox** [comment]
А, лол. Т.е. это тоже самое, что и первое.

> видимо что-то другое. Вот такая ошибка

@h4 ;-)

**Comment by h4** [comment]
@zxqfox вот только у нас нет ни BEMHTML ни BEMTREE. Только bem-core@2.7.0. 

Может сегодня получится полный стектрейс раскопать.
